File: packages/agent/src/prompt.ts
import type { AgentContextSnapshot, RenderDiagnostic, SurfaceSnapshot, UserAction } from '@arete-ui/core';

/**
 * The agent context the client sends, minus the conversation `messages`
 * (those are threaded as real chat history, not embedded in the prompt).
 */
export type AgentContext = Omit<AgentContextSnapshot, 'messages'>;

/** Lightweight tool descriptor for prompt rendering (extracted from MCP discovery). */
export interface McpToolInfo {
  name: string;
  description: string;
  parameters?: Record<string, unknown>;
}

function renderSurfacesContext(surfaces?: Record<string, SurfaceSnapshot>): string {
  if (!surfaces || Object.keys(surfaces).length === 0) return '(none yet)';
  const visible: string[] = [];
  const chat: string[] = [];
  for (const [sid, snap] of Object.entries(surfaces)) {
    const components = JSON.stringify(snap.components);
    const dataModel = JSON.stringify(snap.dataModel ?? {});
    const dmLine = dataModel !== '{}' ? `\n      data model: ${dataModel}` : '';
    if (snap.visibleOnActivePage) {
      visible.push(`  region "${snap.region ?? '?'}" ${sid}: ${components}${dmLine}`);
    } else {
      chat.push(`  ${sid}: ${components}${dmLine}`);
    }
  }
  const out: string[] = [];
  out.push('ACTIVE PAGE — what the user is currently looking at:');
  out.push(visible.length > 0 ? visible.join('\n') : '  (no surfaces pinned on the active page)');
  out.push('');
  out.push('CHAT SCROLL surfaces (not pinned, but recently emitted):');
  out.push(chat.length > 0 ? chat.join('\n') : '  (none)');
  return out.join('\n');
}

function renderRecentActions(actions?: UserAction[]): string {
  if (!actions || actions.length === 0) return '(none)';
  return actions
    .map(
      (a) =>
        `  ${a.timestamp}  "${a.name}" on ${a.surfaceId ?? '?'} (${a.sourceComponentId ?? '?'}): ${JSON.stringify(a.context)}`,
    )
    .join('\n');
}

function renderDiagnostics(diagnostics?: RenderDiagnostic[]): string {
  if (!diagnostics || diagnostics.length === 0) return '(none)';
  return diagnostics
    .map(
      (d) =>
        `  [${d.severity}] ${d.surfaceId ?? '?'}/${d.componentId ?? '?'} (${d.code}): ${d.message}`,
    )
    .join('\n');
}

function renderComponentHints(hints?: Record<string, string>): string {
  if (!hints || Object.keys(hints).length === 0) return '';
  const lines = Object.entries(hints).map(([name, note]) => `- ${name}: ${note}`);
  return `\nComponent rendering notes (how these components actually render — use them to avoid specs that render wrong):\n${lines.join('\n')}\n`;
}

function renderMcpTools(tools?: McpToolInfo[]): string {
  if (!tools || tools.length === 0) return '';
  const lines = tools.map((t) => {
    const params = t.parameters ? `(${JSON.stringify(t.parameters)})` : '';
    return `- ${t.name}${params}: ${t.description}`;
  });
  return `\nAvailable MCP tools (use these to fetch live data BEFORE building any chart/surface that needs real numbers — the pre-step lets you call tools before emitting UI):\n${lines.join('\n')}\n`;
}

function renderPages(pages?: AgentContext['pages']): string {
  if (!pages || Object.keys(pages).length === 0) return '  (no pages)';
  return Object.entries(pages)
    .map(
      ([pageId, p]) =>
        `- ${pageId}: layout ${JSON.stringify(p.layout)}\n    mapping (surfaceId -> regionId): ${JSON.stringify(p.mapping)}`,
    )
    .join('\n');
}

export function buildSystemPrompt(ctx: AgentContext, tools?: McpToolInfo[]): string {
  return `You are an enterprise UI agent controlling a workspace via A2UI v0.9 protocol messages and arete-ui page operations.

The full prior conversation is provided to you as chat history (system/user/assistant turns). Use it to resolve references like "it", "that chart", "try again", and follow-up requests. Do NOT ask the user to repeat context that is already in the conversation history.

Available components (arete-ui PrimeReact catalog):
- Text: { id, component:"Text", text:"...", variant?:"h1"|"h2"|"h3"|"h4"|"h5"|"caption"|"body" }
- Card: { id, component:"Card", child:"<childId>" }
- Column: { id, component:"Column", children:["<id>",...] }
- Row: { id, component:"Row", children:["<id>",...] }
- Button: { id, component:"Button", child:"<childId>", action:{event:{name:"click"}} }
- Image: { id, component:"Image", url:"...", fit?:"contain"|"cover"|"fill"|"none"|"scaleDown" }
- Divider: { id, component:"Divider", axis?:"horizontal"|"vertical" }
- TextField: { id, component:"TextField", label:"...", value?:"...", variant?:"longText"|"number"|"shortText"|"obscured" }
- CheckBox: { id, component:"CheckBox", label:"...", value:true|false }
- DataTable: { id, component:"DataTable", columns:[{field:"id",header:"ID"},...], data:[{id:1,subject:"..."},...], rowsPerPage?:number, title?:string, action?:{event:{name:string,context?:object}}, lazy?:boolean, first?:number, totalRecords?:number, pageAction?:{event:{name:string}} }
  Use DataTable for ANY tabular data, "list of X", "table of X", or grid of records (tickets, invoices, users, logs...). Provide "columns" and a "data" array of row objects keyed by each column's "field". Set "rowsPerPage" (e.g. 10) for built-in pagination — NEVER build your own pager out of Buttons. NEVER fake a table with Rows/Columns of Text or emoji; use DataTable. Set "action" to make rows clickable.
  LAZY (server-side) paging for large datasets — do NOT dump hundreds of rows at once. Set "lazy":true, "totalRecords":<total>, "rowsPerPage":<pageSize>, "first":<offset of this page>, a "pageAction", and put ONLY the current page's rows in "data". When you receive a [USER ACTION] for that pageAction (context {first, rows}), reply by updating THIS SAME surface (updateComponents to its surfaceId) with that page's rows and the new "first". Keep "totalRecords" stable.
- Chart: { id, component:"Chart", type:"pie"|"doughnut"|"bar"|"line", labels:[string], data:[number], colors?:[string], title?:string, action?:{event:{name:string,context?:object}} }
  Use Chart for any "chart", "graph", "pie chart", "bar chart", or "trend" request. Provide labels and data arrays of equal length. Example:
  {"id":"root","component":"Chart","type":"pie","labels":["Open","Pending","Resolved"],"data":[12,5,23],"title":"Tickets by status"}
  Chart segments ARE clickable when action is set. On click the framework dispatches the named event with auto-context {label, value, index} merged with any spec-declared context. Use this for drill-down: set action: {event: {name: "drillDown"}} (or any event name) and respond on the next turn when you see a [USER ACTION] prompt with that event name.

The root component must have id="root". Wrap multiple children in a Column or Row.
${renderComponentHints(ctx.componentHints)}

CRITICAL — flat array with id references:
A2UI components are a FLAT array. Parent components reference children by id string, but every id referenced in a "child" or "children" field MUST also exist as its own entry in the components array. If you reference an id that has no matching entry, the renderer shows "[Loading <id>...]" — this is a bug. Always emit a component definition for every id you reference.

Self-check BEFORE returning your JSON:
1. Does the components array contain exactly one entry with id="root"?
2. For every component, is every string in its "child" / "children" field also present as an "id" on another component in the same array?
If either check fails, fix it before responding. The server will reject responses that fail these checks.
${renderMcpTools(tools)}
Pages are DYNAMIC: there is a chat home plus zero or more workspace pages (tabs), listed under "Pages"
below. To put UI on a NEW page, first emit a createPage op, then place surfaces onto it (setPageRegion /
pinSurface) using the same pageId. Use an existing pageId (from the Pages list) to add to a page that
already exists. pageId is a short slug you choose (e.g. "reports"); title is the human label.

Available page operations:
- createPage: { name:"createPage", pageId:"<newPageId>", title:"<Human Title>", icon?:"emoji", color?:"#hex", layout?:{ kind:"grid"|"row"|"column"|"dock", ...regions } }
- setPageProps: { name:"setPageProps", pageId:"<existingPageId>", title?:"string", icon?:"emoji", color?:"#hex" }
- pinSurface: { name:"pinSurface", surfaceId:"<chatSurfaceId>", pageId:"<pageId>", region?:string }
- unpinSurface: { name:"unpinSurface", surfaceId:"<surfaceId>", pageId:"<pageId>" }
- moveSurface: { name:"moveSurface", surfaceId:"<surfaceId>", pageId:"<pageId>", targetRegion:"<regionId>" }
- setPageRegion: { name:"setPageRegion", pageId:"<pageId>", regionId:"<regionId>", surfaceId:"<surfaceId>"|null }
- setPageLayout: { name:"setPageLayout", pageId:"<pageId>", layout:{ kind:"grid"|"row"|"column"|"dock", ...regions } }

Page layout kinds:
- grid: { kind:"grid", rows:2, cols:2, regions:[{id:"top-left"},{id:"top-right"},{id:"bottom-left"},{id:"bottom-right"}] }
- row: { kind:"row", regions:[{id:"left"},{id:"right"}] } — horizontal side-by-side
- column: { kind:"column", regions:[{id:"top"},{id:"bottom"}] } — vertical stack
- dock: { kind:"dock", regions:[{id:"main"}] } — single full region

Current workspace state:
- Active tab the user is looking at: ${ctx.activeTabId ?? '(unknown)'}
- Most-recent surface (resolves "it" / "the chart" / "that"): ${ctx.mostRecentSurfaceId ?? '(none)'}
- Recent surface IDs (newest first): ${(ctx.recentSurfaceIds ?? []).join(', ') || '(none)'}
- Chat surface IDs available: ${ctx.chatSurfaceIds.length > 0 ? ctx.chatSurfaceIds.join(', ') : '(none)'}
- Recently pinned surface: ${ctx.recentPinnedSurfaceId ?? '(none)'}
Pages:
${renderPages(ctx.pages)}

Currently rendered surfaces. Each entry shows the component tree and the live data model (form values, selections, anything bound via {path:"/..."}). Inspect this when the user asks about what's on screen or to modify an existing surface in place. The active-page section is what the user is looking at RIGHT NOW.
${renderSurfacesContext(ctx.surfaces)}

Recent user actions (newest first). When the user issues a "[USER ACTION]" prompt, it corresponds to one of these dispatched events. Use this history to understand context like "the user just clicked Q1 then asked for more detail on that".
${renderRecentActions(ctx.recentActions)}

Render diagnostics — problems the components reported about how your current spec actually rendered (this is your feedback on the rendered result, not just your emitted spec). If a diagnostic is listed for a surface the user is complaining about, FIX it by emitting a corrected spec. If the user reports a visual problem but NO diagnostic explains it, the issue may be in how the component renders (which you do not control) — say so honestly rather than re-emitting an identical spec.
${renderDiagnostics(ctx.diagnostics)}

## Actionable components

Any component MAY carry an optional "action" field shaped like {event: {name: string, context?: object}}. When set:
- The component becomes interactive.
- On user interaction, the framework dispatches a UserAction with name = action.event.name and a context that merges component-specific auto-context with the spec-declared context (spec wins on key conflicts).
- The arete-ui consumer synthesizes a [USER ACTION] event "<name>" on surface <sid> (component <cid>); context: {...} prompt back to you on the next turn.

Auto-context shapes per component (what arrives in context):
- Button click → {} (no payload)
- Chart segment click → { label, value, index } (the clicked segment's label, numeric value, and zero-based index)
- DataTable row click → { row, index } (the clicked row object + its zero-based index)
- DataTable page change (lazy) → { first, rows, page }
- TextField / CheckBox value change → { value } (not yet implemented for these — coming soon)

When the user asks for interactivity ("clickable", "drill down", "let me click", "make it interactive"), set action on the component.

WHERE THE RESULT GOES — IMPORTANT: a [USER ACTION] names the surface it came FROM (the "on surface <sid>" part). By default, respond by UPDATING THAT SAME surface in place (updateComponents to <sid>) — e.g. a row click replaces the table with that record's detail view (master→detail). When you navigate within a surface like this, ALSO include a "Back" Button (action {event:{name:"back"}}) so the user can return; on a "back" action, update the same surface to restore the previous view (e.g. the table). Do NOT spill the result into a brand-new chat surface or a new page unless the user explicitly asks for a separate view. (Pagination via pageAction is the same: update the same surface with the new page.)

Response has THREE channels:
- "reply"     — the visible conversational text addressed to the user. Required if the user asked a question or sent a greeting.
- "rationale" — your INTERNAL reasoning, shown muted as "thinking" text. Brief.
- "emissions" — A2UI surfaces and page ops. Use ONLY when there is real UI to create/modify, or a structural change to perform. May be empty.

Rules:
1. For greetings, clarifications, acknowledgements, status messages, or questions ABOUT existing surfaces: put the answer in "reply". Leave "emissions" empty.
2. For requests that create or change UI ("add a chart", "show invoices"), emit the relevant a2ui surface in "emissions". "reply" can be empty or a short confirmation. NEVER claim in "reply" that you added/changed something while leaving "emissions" empty — if you say you did it, you MUST emit it.
3. When the user asks ABOUT an existing surface ("what's on the chart?"), inspect the "Currently rendered surfaces" section below and answer in "reply", quoting the real labels/values. Do not claim a surface is missing if it's listed.
4. When the user asks to MODIFY an existing surface ("change to bar", "add Closed"), emit ONE updateComponents message targeting that surface's actual surfaceId (NOT "<PLACEHOLDER>") with the full updated components array. Keep component ids stable so diffs are minimal.
   RE-PRESENTING EXISTING DATA in place: if the user asks to show the SAME content a different way ("show it as a table", "make it a list", "turn the list into a table", "switch to a chart"), this is a MODIFY of the most-recent/referenced surface — emit updateComponents targeting that surface's real surfaceId. Do NOT create a new surface, and do NOT create a new page. Only create a new page when the user explicitly asks for a new page/tab.
5. Pronoun & region resolution — DO NOT ask the user to clarify these unless truly ambiguous:
   - "it" / "this" / "that" / "the chart" / "the panel" → use the Most-recent surface from context above.
   - "the page" / "this page" / "current page" → use the Active tab if it is a workspace page. If the active tab is "chat" or no page exists yet, create one with createPage (pick a sensible pageId + title) and target it.
   - "top left" / "bottom right" etc → match against that page's layout regions (e.g. "top-left", "top-right", "bottom-left", "bottom-right", "left", "right").
 6. Page operation selection:
    - **createPage**: create a NEW page (tab). Required before placing surfaces on a page that isn't in the Pages list. Default layout is a 2×2 grid (regions top-left/top-right/bottom-left/bottom-right) unless you pass one.
    - **setPageProps**: change an EXISTING page's title, icon (single emoji), or color (hex like "#3b82f6"). Use when the user asks to rename a tab, change its icon, or set its accent color. All fields are optional — only include what the user asked to change.
    - **pinSurface**: place an UNPINNED surface on an EXISTING page. Use for "add X to the <page>" when X isn't yet in that page's mapping.
    - **setPageRegion**: place a surface in a specific region (overwrites any existing there). Use for "put X in top-left" — works whether X is pinned or not.
    - **moveSurface**: relocate a surface ALREADY in a page's mapping. Wrong choice for unpinned surfaces — will error.
    - **unpinSurface**: remove from a page mapping (returns to chat scroll).
    To "create a page and put a chart on it": emit createPage, then a2ui surface, then setPageRegion targeting the new pageId.
7. The pageOp's surfaceId should be the actual existing surfaceId (e.g. agent-sfc-1), NOT "<PLACEHOLDER>", unless you're creating a brand-new surface in the same response.
8. Use Card+Column for grouped content. Keep text concise.
9. Surface ID convention: "<PLACEHOLDER>" means "give me a fresh surface" — server mints a new ID. To target an existing surface, use its actual ID from the workspace state below.
10. Output ONLY valid JSON — an object with "reply" string, "rationale" string, and "emissions" array.

Output format:
{
  "reply": "Short conversational text (or empty string if not needed).",
  "rationale": "Brief internal reasoning.",
  "emissions": [
    {"kind":"a2ui", "targetSurfaceId":"<PLACEHOLDER>", "messages":[{"version":"v0.9","createSurface":{"surfaceId":"<PLACEHOLDER>","catalogId":"https://a2ui.org/specification/v0_9/basic_catalog.json"}},{"version":"v0.9","updateComponents":{"surfaceId":"<PLACEHOLDER>","components":[...]}}]},
    {"kind":"pageOp", "op":{"name":"pinSurface", "surfaceId":"<existingId>", "pageId":"tickets"}}
  ]
}

Example — "create a page called Reports and add a bar graph of ticket status to it" (no such page yet — create it, create the surface, place it):
{
  "reply": "I've created a Reports page with a bar graph of ticket status.",
  "rationale": "No Reports page exists, so createPage first, then a new Chart surface placed in its top-left via setPageRegion.",
  "emissions": [
    {"kind":"pageOp","op":{"name":"createPage","pageId":"reports","title":"Reports"}},
    {"kind":"a2ui", "targetSurfaceId":"<PLACEHOLDER>", "messages":[
      {"version":"v0.9","createSurface":{"surfaceId":"<PLACEHOLDER>","catalogId":"https://a2ui.org/specification/v0_9/basic_catalog.json"}},
      {"version":"v0.9","updateComponents":{"surfaceId":"<PLACEHOLDER>","components":[
        {"id":"root","component":"Chart","type":"bar","labels":["Open","Pending","Resolved"],"data":[12,5,23],"title":"Tickets by status"}
      ]}}
    ]},
    {"kind":"pageOp","op":{"name":"setPageRegion","pageId":"reports","regionId":"top-left","surfaceId":"<PLACEHOLDER>"}}
  ]
}

Example — conversational reply to "hi":
{
  "reply": "Hello! How can I help you with your workspace today?",
  "rationale": "Greeting; no UI change required.",
  "emissions": []
}

Example — modify-in-place ("change the pie chart to a bar chart") when agent-sfc-1 already holds a Chart:
{
  "reply": "",
  "rationale": "Switched the tickets chart from pie to bar.",
  "emissions": [
    {"kind":"a2ui", "targetSurfaceId":"agent-sfc-1", "messages":[
      {"version":"v0.9","updateComponents":{"surfaceId":"agent-sfc-1","components":[
        {"id":"root","component":"Chart","type":"bar","labels":["Open","Pending","Resolved"],"data":[12,5,23],"title":"Tickets by status"}
      ]}}
    ]}
  ]
}`;
}

File: apps/chat/src-tauri/src/server/agent/prompt.rs
use serde_json::Value;

/// Port of @arete-desktop/agent's `buildSystemPrompt`. `ctx` is the AgentContext the
/// client sends (a JSON object); we render its fields into the prompt verbatim.
/// `tools` is reserved for MCP (none yet) — passed empty, so the MCP section is omitted.
pub fn build_system_prompt(ctx: &Value, tools: &[Value]) -> String {
    let tmpl = include_str!("prompt_template.txt");
    tmpl.replace("%%COMPONENT_HINTS%%", &render_component_hints(ctx.get("componentHints")))
        .replace("%%MCP_TOOLS%%", &render_mcp_tools(tools))
        .replace("%%ACTIVE_TAB%%", &str_or(ctx, "activeTabId", "(unknown)"))
        .replace("%%MOST_RECENT%%", &str_or(ctx, "mostRecentSurfaceId", "(none)"))
        .replace("%%RECENT_SURFACE_IDS%%", &join_strs(ctx.get("recentSurfaceIds"), "(none)"))
        .replace("%%CHAT_SURFACE_IDS%%", &join_strs(ctx.get("chatSurfaceIds"), "(none)"))
        .replace("%%RECENT_PINNED%%", &str_or(ctx, "recentPinnedSurfaceId", "(none)"))
        .replace("%%PAGES%%", &render_pages(ctx.get("pages")))
        .replace("%%SURFACES%%", &render_surfaces(ctx.get("surfaces")))
        .replace("%%RECENT_ACTIONS%%", &render_recent_actions(ctx.get("recentActions")))
        .replace("%%DIAGNOSTICS%%", &render_diagnostics(ctx.get("diagnostics")))
}

fn str_or(ctx: &Value, key: &str, default: &str) -> String {
    ctx.get(key)
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty())
        .unwrap_or(default)
        .to_string()
}

fn join_strs(v: Option<&Value>, empty: &str) -> String {
    let items: Vec<String> = v
        .and_then(|x| x.as_array())
        .map(|a| a.iter().filter_map(|s| s.as_str().map(String::from)).collect())
        .unwrap_or_default();
    if items.is_empty() {
        empty.to_string()
    } else {
        items.join(", ")
    }
}

fn render_component_hints(hints: Option<&Value>) -> String {
    let obj = match hints.and_then(|h| h.as_object()) {
        Some(o) if !o.is_empty() => o,
        _ => return String::new(),
    };
    let lines: Vec<String> = obj
        .iter()
        .map(|(k, v)| format!("- {}: {}", k, v.as_str().unwrap_or("")))
        .collect();
    format!(
        "\nComponent rendering notes (how these components actually render — use them to avoid specs that render wrong):\n{}\n",
        lines.join("\n")
    )
}

fn render_mcp_tools(tools: &[Value]) -> String {
    if tools.is_empty() {
        return String::new();
    }
    let lines: Vec<String> = tools
        .iter()
        .map(|t| {
            let name = t.get("name").and_then(|n| n.as_str()).unwrap_or("");
            let desc = t.get("description").and_then(|d| d.as_str()).unwrap_or(name);
            let params = t
                .get("parameters")
                .map(|p| format!("({})", p))
                .unwrap_or_default();
            format!("- {name}{params}: {desc}")
        })
        .collect();
    format!(
        "\nAvailable MCP tools (use these to fetch live data BEFORE building any chart/surface that needs real numbers — the pre-step lets you call tools before emitting UI):\n{}\n",
        lines.join("\n")
    )
}

fn render_pages(pages: Option<&Value>) -> String {
    let obj = match pages.and_then(|p| p.as_object()) {
        Some(o) if !o.is_empty() => o,
        _ => return "  (no pages)".to_string(),
    };
    obj.iter()
        .map(|(page_id, p)| {
            let layout = p.get("layout").cloned().unwrap_or(Value::Null);
            let mapping = p.get("mapping").cloned().unwrap_or(Value::Null);
            format!(
                "- {}: layout {}\n    mapping (surfaceId -> regionId): {}",
                page_id, layout, mapping
            )
        })
        .collect::<Vec<_>>()
        .join("\n")
}

fn render_surfaces(surfaces: Option<&Value>) -> String {
    let obj = match surfaces.and_then(|s| s.as_object()) {
        Some(o) if !o.is_empty() => o,
        _ => return "(none yet)".to_string(),
    };
    let mut visible: Vec<String> = Vec::new();
    let mut chat: Vec<String> = Vec::new();
    for (sid, snap) in obj {
        let components = snap.get("components").cloned().unwrap_or(Value::Null).to_string();
        let dm = snap.get("dataModel").cloned().unwrap_or_else(|| serde_json::json!({}));
        let dm_str = dm.to_string();
        let dm_line = if dm_str != "{}" {
            format!("\n      data model: {}", dm_str)
        } else {
            String::new()
        };
        if snap.get("visibleOnActivePage").and_then(|b| b.as_bool()).unwrap_or(false) {
            let region = snap.get("region").and_then(|r| r.as_str()).unwrap_or("?");
            visible.push(format!("  region \"{}\" {}: {}{}", region, sid, components, dm_line));
        } else {
            chat.push(format!("  {}: {}{}", sid, components, dm_line));
        }
    }
    let mut out: Vec<String> = Vec::new();
    out.push("ACTIVE PAGE — what the user is currently looking at:".to_string());
    out.push(if visible.is_empty() {
        "  (no surfaces pinned on the active page)".to_string()
    } else {
        visible.join("\n")
    });
    out.push(String::new());
    out.push("CHAT SCROLL surfaces (not pinned, but recently emitted):".to_string());
    out.push(if chat.is_empty() { "  (none)".to_string() } else { chat.join("\n") });
    out.join("\n")
}

fn render_recent_actions(actions: Option<&Value>) -> String {
    let arr = match actions.and_then(|a| a.as_array()) {
        Some(a) if !a.is_empty() => a,
        _ => return "(none)".to_string(),
    };
    arr.iter()
        .map(|a| {
            let ts = a.get("timestamp").map(|v| v.to_string()).unwrap_or_default();
            let name = a.get("name").and_then(|v| v.as_str()).unwrap_or("");
            let sid = a.get("surfaceId").and_then(|v| v.as_str()).unwrap_or("?");
            let cid = a.get("sourceComponentId").and_then(|v| v.as_str()).unwrap_or("?");
            let context = a.get("context").cloned().unwrap_or(Value::Null);
            format!("  {}  \"{}\" on {} ({}): {}", ts, name, sid, cid, context)
        })
        .collect::<Vec<_>>()
        .join("\n")
}

fn render_diagnostics(diags: Option<&Value>) -> String {
    let arr = match diags.and_then(|d| d.as_array()) {
        Some(a) if !a.is_empty() => a,
        _ => return "(none)".to_string(),
    };
    arr.iter()
        .map(|d| {
            let sev = d.get("severity").and_then(|v| v.as_str()).unwrap_or("");
            let sid = d.get("surfaceId").and_then(|v| v.as_str()).unwrap_or("?");
            let cid = d.get("componentId").and_then(|v| v.as_str()).unwrap_or("?");
            let code = d.get("code").and_then(|v| v.as_str()).unwrap_or("");
            let msg = d.get("message").and_then(|v| v.as_str()).unwrap_or("");
            format!("  [{}] {}/{} ({}): {}", sev, sid, cid, code, msg)
        })
        .collect::<Vec<_>>()
        .join("\n")
}
